### PR TITLE
DOCSP-19737: Android SDK, Flexible Sync

### PR DESCRIPTION
**Pull request by nathan-contino-mongo**
## Pull Request Info
Strongly structurally inspired by the iOS equivalent PR, which added the "open a realm" example and the "flexible sync" page. Deviations only to accommodate a different SDK design and different usage patterns.

### Jira

- https://jira.mongodb.org/browse/DOCSP-19737

### Staged Changes (Requires MongoDB Corp SSO)

- [Open a Realm - Flexible Sync](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/android-flexible-sync/sdk/android/examples/open-and-close-a-realm/#open-a-synced-realm-with-a-flexible-sync-configuration)
- [Android SDK - Flexible Sync](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/android-flexible-sync/sdk/android/examples/flexible-sync/)
- [Android SDK - Realm Sync](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/android-flexible-sync/sdk/android/fundamentals/realm-sync/)

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sdk/android/examples/flexible-sync.txt [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+   If you subscribe to the same query more than once under different names, 
+   Realm persists both subscriptions to the subscription set.
+
+.. _android-sync-add-subscription:
+
+Add a Subscription
+~~~~~~~~~~~~~~~~~~
+
+Add a subscription in a subscriptions write block. You append each
```
Comment (dacharyc) [inline source/sdk/android/examples/flexible-sync.txt:125]: I'm realizing we don't have anything on the page about which RQL operators are supported in Flex Sync queries. I should add this to the Swift page, but I think somewhere in this section, maybe, we should include this note: [includes/note-unsupported-flex-sync-rql-operators.rst](https://github.com/mongodb/docs-realm/blob/master/source/includes/note-unsupported-flex-sync-rql-operators.rst)

**Review by dacharyc — COMMENTED** [review]
Overall, looks great! Just a handful of comments and suggestions.

Also, we're adding a [Realm Sync](https://docs.mongodb.com/realm/sdk/swift/fundamentals/realm-sync/) page to the `Fundamentals` section - I did it in a separate PR in Swift because I knew there was going to be a delay before the Swift API release, but I think you can do it here in this PR to release it all together for Android.

File: source/sdk/android/examples/open-and-close-a-realm.txt [hunk 2]
```diff
@@ -267,6 +267,41 @@ that you use to :ref:`open a synced realm while online
 pattern, so cross-platform developers should consult the documentation
 for each SDK to learn more.
 
+.. _android-flexible-sync-open-realm:
+
+Open a Synced Realm with a Flexible Sync Configuration
```
[the remark this answers is not in the document]
Comment (nathan-contino-mongo) [reply source/sdk/android/examples/open-and-close-a-realm.txt:272]: In this SDK, both partition-based sync and flexible sync use the same SyncConfiguration interface -- it's just that flexible sync omits the partition value, and adds a single Builder method. I've made it clearer that the first example demonstrates opening a *partition-based* realm, and I'm adding a line about how flexible sync omits the partition value.

**Review by dacharyc — APPROVED** [review]
Nice work - LGTM!

File: examples/android/sync/app/src/androidTest/java/com/mongodb/realm/examples/java/FlexibleSyncTest.java [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+            Credentials credentials = Credentials.anonymous();
+            app.loginAsync(credentials, it -> {
+                if (it.isSuccess()) {
+                    User user = it.get();
+                    SyncConfiguration config = new SyncConfiguration.Builder(app.currentUser())
+                            .initialSubscriptions(new SyncConfiguration.InitialFlexibleSyncSubscriptions() {
+                                @Override
+                                public void configure(Realm realm, MutableSubscriptionSet subscriptions) {
+                                    // :code-block-start: implicitly-named-subscription
+                                    subscriptions.add(Subscription.create(null,
```
Comment (cmelchior) [inline examples/android/sync/app/src/androidTest/java/com/mongodb/realm/examples/java/FlexibleSyncTest.java:141]: There is a specific override for this. You can just call `Subscription.create(query)` instead.

File: source/sdk/android/examples/flexible-sync.txt [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+To use Flexible Sync in an Android client:
+
+- :ref:`Configure Flexible Sync on the backend <enable-flexible-sync>`
+- :ref:`Initialize the app <android-quick-start-init-app>`
+- :ref:`Authenticate a user <android-quick-start-authenticate>` in
+  your client project.
+- :ref:`Open the synced Realm with a Flexible Sync configuration <android-flexible-sync-open-realm>`
+- :ref:`Add subscriptions to the client application <android-sync-subscribe-to-queryable-fields>`
+
+You can add, update, and remove query subscriptions to determine which data
```
Comment (cmelchior) [inline source/sdk/android/examples/flexible-sync.txt:30]: I don't think the term "query subscriptions" has been introduced before, which makes this a bit hard to follow. Maybe try to add a sentence to the first intro line about what a subscription actually is

File: source/sdk/android/examples/flexible-sync.txt [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+          .. literalinclude:: /examples/generated/android/sync/FlexibleSyncTest.codeblock.explicitly-named-subscription.kt
+            :language: kotlin
+
+      .. tab::
+          :tabid: java
+
+          .. literalinclude:: /examples/generated/android/sync/FlexibleSyncTest.codeblock.explicitly-named-subscription.java
+            :language: java
+
+   If you pass ``null`` as the name when creating a subscription, you
```
Comment (cmelchior) [inline source/sdk/android/examples/flexible-sync.txt:93]: Note, there exists a separate constructor if you don't want a name.

**Review by cmelchior — APPROVED** [review]
Nice...only had a few minor comments.

File: source/sdk/android/examples/flexible-sync.txt (hunk 4 above)
Comment (nathan-contino-mongo) [reply source/sdk/android/examples/flexible-sync.txt:30]: We also use and introduce the term on the backend documentation, linked above this. Bit tricky to frontload too much description of "query subscriptions" in this example here -- that reference material is mostly constrained to the backend explanation and the fundamentals page also added in this PR.

File: examples/android/sync/app/src/androidTest/java/com/mongodb/realm/examples/java/FlexibleSyncTest.java (hunk 3 above)
Comment (nathan-contino-mongo) [reply examples/android/sync/app/src/androidTest/java/com/mongodb/realm/examples/java/FlexibleSyncTest.java:141]: Update: turns out it doesn't use "null" as a realmFileName -- instead, it uses "default.realm", which very very easily creates conflicts with non-flexible sync configurations if an app uses both. I would strongly consider generating names off of queries or something like that instead because this seems very likely to trip up users.
